PhysicsTools/NanoAOD/plugins/MuonIDTableProducer.cc

```diff
@@ -0,0 +1,116 @@
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+#include "DataFormats/NanoAOD/interface/FlatTable.h"
+
+
+class MuonIDTableProducer : public edm::global::EDProducer<> {
+    public:
+        explicit MuonIDTableProducer(const edm::ParameterSet &iConfig) :
+            name_(iConfig.getParameter<std::string>("name")),
+            src_(consumes<std::vector<pat::Muon>>(iConfig.getParameter<edm::InputTag>("muons"))),
+            srcVtx_(consumes<std::vector<reco::Vertex>>(iConfig.getParameter<edm::InputTag>("vertices")))
+        {
+            produces<FlatTable>();
+        }
+
+        ~MuonIDTableProducer() override {};
+
+        static void fillDescriptions(edm::ConfigurationDescriptions & descriptions) {
+            edm::ParameterSetDescription desc;
+            desc.add<edm::InputTag>("muons")->setComment("input muon collection");
+            desc.add<edm::InputTag>("vertices", edm::InputTag("offlineSlimmedPrimaryVertices"))->setComment("input vertex collection, for dxy/dz");
+            desc.add<std::string>("name")->setComment("name of the muon FlatTable we are extending with IDs");
+            descriptions.add("muonIDTable", desc);
+        }
+
+    private:
+        void produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const override ;
+
+        std::string name_;
+        edm::EDGetTokenT<std::vector<pat::Muon>> src_;
+        edm::EDGetTokenT<std::vector<reco::Vertex>> srcVtx_;
+
+        static bool isMediumMuonHIP(const pat::Muon& muon) ;
+        static bool isSoftMuonHIP(const pat::Muon& muon, const reco::Vertex& vtx) ;
+        static bool isTrackerHighPt(const pat::Muon & mu, const reco::Vertex & primaryVertex) ;
+};
+
+// ------------ method called to produce the data  ------------
+void
+MuonIDTableProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const 
+{
+
+    edm::Handle<std::vector<pat::Muon>> muons;
+    iEvent.getByToken(src_, muons);
+    edm::Handle<std::vector<reco::Vertex>> vertices;
+    iEvent.getByToken(srcVtx_, vertices);
+
+    unsigned int ncand = muons->size();
+
+    const reco::Vertex & pv = vertices->front(); // consistent with IP information in slimmedLeptons.
+
+    bool isRun2016BCDEF = (272007 <= iEvent.run() && iEvent.run() <= 278808);
+    std::vector<uint8_t> tight(ncand, 0), highPt(ncand, 0), soft(ncand, 0), medium(ncand, 0);
+    for (unsigned int i = 0; i < ncand; ++i) {
+        const pat::Muon & mu = (*muons)[i];
+        tight[i] = muon::isTightMuon(mu, pv);
+        highPt[i] = muon::isHighPtMuon(mu, pv) ? 2 : isTrackerHighPt(mu,pv);
+        soft[i] = isRun2016BCDEF ? isSoftMuonHIP(mu,pv) : muon::isTightMuon(mu, pv);
+        medium[i] = isRun2016BCDEF ? isMediumMuonHIP(mu) : muon::isMediumMuon(mu);
```

Comment:
this overlaps with #20612 
and should be updated once that is merged

Reply:
this was coded while waiting for that PR to appear. it will be removed when the PR will be merged and available in a prerelease.

Reply:
#20612 was merged already